x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java

```diff
@@ -222,64 +234,151 @@ private void filterResults(String javaRegex, GetAliasesResponse aliases, GetInde
+            return IndexResolution.notFound(indexPattern);
+        }
+
+        StringBuilder errorMessage = new StringBuilder();
+
+        NavigableSet<Entry<String, Map<String, FieldCapabilities>>> sortedFields = new TreeSet<>(
+                // for some reason .reversed doesn't work (prolly due to inference)
+                Collections.reverseOrder(Comparator.comparing(Entry::getKey)));
+        sortedFields.addAll(fieldCaps.entrySet());
+
+        Map<String, EsField> hierarchicalMapping = new TreeMap<>();
+        Map<String, EsField> flattedMapping = new LinkedHashMap<>();
+        
+        // sort keys descending in order to easily detect multi-fields (a.b.c multi-field of a.b)
+        // without sorting, they can still be detected however without the emptyMap optimization
+        // (fields without multi-fields have no children)
+        for (Entry<String, Map<String, FieldCapabilities>> entry : sortedFields) {
+            String name = entry.getKey();
+            // skip internal fields
+            if (!name.startsWith("_")) {
+                Map<String, FieldCapabilities> types = entry.getValue();
+                // field is mapped differently across indices
+                if (types.size() > 1) {
+                    // build error message
+                    for (Entry<String, FieldCapabilities> type : types.entrySet()) {
+                        if (errorMessage.length() > 0) {
+                            errorMessage.append(", ");
+                        }
+                        errorMessage.append("[");
+                        errorMessage.append(type.getKey());
+                        errorMessage.append("] in ");
+                        errorMessage.append(Arrays.toString(type.getValue().indices()));
                     }
+
+                    errorMessage.insert(0,
+                            "[" + indexPattern + "] points to indices with incompatible mappings; " +
+                            "field [" + name + "] is mapped in [" + types.size() + "] different ways: ");
+                }
+                if (errorMessage.length() > 0) {
+                    return IndexResolution.invalid(errorMessage.toString());
+                }
+                
+                FieldCapabilities fieldCap = types.values().iterator().next();
+                // validate search/agg-able
+                if (fieldCap.isAggregatable() && fieldCap.nonAggregatableIndices() != null) {
+                    errorMessage.append("[" + indexPattern + "] points to indices with incompatible mappings: ");
+                    errorMessage.append("field [" + name + "] is aggregateable except in ");
+                    errorMessage.append(Arrays.toString(fieldCap.nonAggregatableIndices()));
+                }
+                if (fieldCap.isSearchable() && fieldCap.nonSearchableIndices() != null) {
+                    errorMessage.append("[" + indexPattern + "] points to indices with incompatible mappings: ");
```

Comment:
is this check mutually exclusive with the above? If yes I would prefer an `else if` to denote that, otherwise the `errorMessage` should be checked and a `, ` should be appended first.

Reply:
Unfortunately, the two are not exclusive. I'll add a `,` in.